consensus/scheme/rolldpos/fsmevent.go

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) 2018 IoTeX
+// This is an alpha (internal) release and is not suitable for production. This source code is provided 'as is' and no
+// warranties are given as to title or non-infringement, merchantability or fitness for purpose and, to the extent
+// permitted by law, all liability for your use of the code is disclaimed. This source code is governed by Apache
+// License 2.0 that can be found in the LICENSE file.
+
+package rolldpos
+
+import (
+	"time"
+
+	"github.com/facebookgo/clock"
+	fsm "github.com/zjshen14/go-fsm"
+
+	"github.com/iotexproject/iotex-core/blockchain"
+	"github.com/iotexproject/iotex-core/endorsement"
+	"github.com/iotexproject/iotex-core/iotxaddress"
+	"github.com/iotexproject/iotex-core/pkg/hash"
+	"github.com/iotexproject/iotex-core/proto"
+)
+
+// iConsensusEvt is the interface of all events for the consensusEvt FSM
+type iConsensusEvt interface {
+	fsm.Event
+	height() uint64
+	round() uint32
+	timestamp() time.Time
+	// TODO: we need to add height or some other ctx to identify which consensus round the event is associated to
+}
+
+type consensusEvt struct {
+	h  uint64
+	r  uint32
+	t  fsm.EventType
+	ts time.Time
+}
+
+func newCEvt(t fsm.EventType, height uint64, round uint32, c clock.Clock) *consensusEvt {
+	return &consensusEvt{
+		h:  height,
+		r:  round,
+		t:  t,
+		ts: c.Now(),
+	}
+}
+
+func (e *consensusEvt) height() uint64 { return e.h }
+
+func (e *consensusEvt) round() uint32 { return e.r }
+
+func (e *consensusEvt) Type() fsm.EventType { return e.t }
+
+func (e *consensusEvt) timestamp() time.Time { return e.ts }
+
+type proposeBlkEvt struct {
+	consensusEvt
+	block *blockchain.Block
+}
+
+func newProposeBlkEvt(block *blockchain.Block, round uint32, c clock.Clock) *proposeBlkEvt {
+	return &proposeBlkEvt{
+		consensusEvt: *newCEvt(eProposeBlock, block.Height(), round, c),
```

Comment:
define consensusEvt as *consensusEvt? then no need to dereference here, or let newCEvt() to directly return the struct?

Reply:
Will refactor the consensus event related code. Focusing on the logic part first